controller/events.go

```diff
@@ -0,0 +1,240 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/gopkg.in/inconshreveable/log15.v2"
+	ct "github.com/flynn/flynn/controller/types"
+)
+
+// EventQueue is a FIFO queue for storing app events (used by EventSubscriber
+// so that slow receivers don't block the main EventListener loop).
+type EventQueue struct {
+	events []*ct.AppEvent
+	cond   *sync.Cond
+	closed bool
+}
+
+func newEventQueue() *EventQueue {
+	return &EventQueue{cond: sync.NewCond(&sync.Mutex{})}
+}
+
+// Push pushes an event onto the tail of the queue.
+func (e *EventQueue) Push(event *ct.AppEvent) {
+	e.cond.L.Lock()
+	e.events = append(e.events, event)
+	e.cond.L.Unlock()
+	e.cond.Signal()
+}
+
+// Pop removes and returns an event from the head of the queue, returning
+// nil if the queue is closed.
+func (e *EventQueue) Pop() *ct.AppEvent {
+	e.cond.L.Lock()
+	defer e.cond.L.Unlock()
+	if e.closed {
+		return nil
+	}
+	for len(e.events) == 0 {
+		e.cond.Wait()
+		if e.closed {
+			return nil
+		}
+	}
+	event := e.events[0]
+	e.events = e.events[1:]
```

Comment:
I think this will result in excessive copying, as the pointer will always be headed towards the end of the backing array. A better approach would be a ring buffer (which would allow specifying a limit on the buffer size easily) or using `container/list` which implements a linked list.